hasktorch/test/StaticSpec.hs

```diff
@@ -33,32 +33,68 @@ import Torch.Static.Native
 import qualified Torch.TensorOptions as D
 import Data.Reflection
 
+checkDynamicTensorAttributes
+  :: forall dtype shape
+    . (TensorOptions dtype shape)
+  => Tensor dtype shape
+  -> IO ()
+checkDynamicTensorAttributes t = do
+  shape t `shouldBe` optionsRuntimeShape @dtype @shape
+  dtype t `shouldBe` optionsRuntimeDType @dtype @shape
+
 spec :: Spec
 spec = do
   it "sin" $ do
     let t = sin zeros :: Tensor Float '[2,3]
-    D.shape (toDynamic t) `shouldBe` [2,3]
-    D.dtype (toDynamic t) `shouldBe` D.Float
+    checkDynamicTensorAttributes t
   it "ones" $ do
     let t = ones :: Tensor Float '[2,3]
-    D.shape (toDynamic t) `shouldBe` [2,3]
-    D.dtype (toDynamic t) `shouldBe` D.Float
+    checkDynamicTensorAttributes t
   it "zeros" $ do
     let t = zeros :: Tensor Float '[2,3]
-    D.shape (toDynamic t) `shouldBe` [2,3]
-    D.dtype (toDynamic t) `shouldBe` D.Float
+    checkDynamicTensorAttributes t
   it "zeros with double" $ do
     let t = zeros :: Tensor Double '[2,3]
-    D.shape (toDynamic t) `shouldBe` [2,3]
-    D.dtype (toDynamic t) `shouldBe` D.Double
+    checkDynamicTensorAttributes t
   it "randn" $ do
     t <- randn :: IO (Tensor Double '[2,3])
-    D.shape (toDynamic t) `shouldBe` [2,3]
-    D.dtype (toDynamic t) `shouldBe` D.Double
-  it "add" $ do
-    let a = ones :: Tensor Float '[2,3]
-    let b = ones :: Tensor Float '[2,3]
-    let c = a + b :: Tensor Float '[2,3]
-    D.asValue (toDynamic c) `shouldBe` ([[2,2,2],[2,2,2]]::[[Float]])
-    D.shape (toDynamic c) `shouldBe` [2,3]
-    D.dtype (toDynamic c) `shouldBe` D.Float
+    checkDynamicTensorAttributes t
+  describe "add" $ do
+    it "works on tensors of identical shapes" $ do
+      let a = ones :: Tensor Float '[2, 3]
+      let b = ones :: Tensor Float '[2, 3]
+      let c = add a b
+      checkDynamicTensorAttributes c
+      D.asValue (toDynamic c) `shouldBe` ([[2, 2, 2], [2, 2, 2]] :: [[Float]])
+    it "works on broadcastable tensors of different shapes" $ do
+      let a = ones :: Tensor Float '[3, 1, 4, 1]
+      let b = ones :: Tensor Float '[2, 1, 1]
+      let c = add a b
+      checkDynamicTensorAttributes c
+      D.asValue (toDynamic c) `shouldBe` ([[2, 2], [2, 2], [2, 2]] :: [[Float]])
```

Comment:
I feel this is something storage.
> D.asValue (toDynamic c) `shouldBe` ([[2, 2], [2, 2], [2, 2]] :: [[Float]])
c's shape is  [3,2,4,1].

```
> add (ones' [3,1,4,1]) (ones' [2,1,1])
Tensor Float [3,2,4,1]
```

I think this becomes as follows.

```
D.asValue (toDynamic c) `shouldBe` ([[[[2.0],[2.0],[2.0],[2.0]],[[2.0],[2.0],[2.0],[2.0]]],[[[2.0],[2.0],[2.0],[2.0]],[[2.0],[2.0],[2.0],[2.0]]],[[[2.0],[2.0],[2.0],[2.0]],[[2.0],[2.0],[2.0],[2.0]]]] :: [[[[Float]]]])
 
```

Reply:
@junjihashimoto I agree, and I don't understand how this test could have passed. I just added the correct expectations while leaving the old (wrong ones). Tests are passing, so I don't understand what's going on.

it seems like `D.asValue (toDynamic c) :: [[Float]]` is different from `D.asValue (toDynamic c) :: [[[[Float]]]]` for the same `c`

Reply:
We need the validation of asValue to check the shape.